### ISSUE TITLE
[Python][CI] Nightly pandas/numpy builds are crashing

### ISSUE THREAD

**Issue by jorisvandenbossche** [apache/arrow#40364]
The last run of the pandas nightly/upstream_devel builds have been segfaulting. The run of 2024-03-01 was still working (one actual error in the pandas nightly, green in upstream_devel). Now both are crashing, also after some other tests failing.

Example: https://github.com/ursacomputing/crossbow/actions/runs/8134146382/job/22226672519

My suspicion is that this is more likely related to numpy main (they have been merging a lot of changes the last days ramping up to branching 2.0)

**Comment by jorisvandenbossche** [comment]
Hmm, it seems the crashes were in the one but last build, the ones from last night are failing "normally" again. Maybe there was a brief incompatibility between the numpy and pandas nightly wheels, or just something on numpy's side that was fixed in the meantime.

Anyway, we still get a bunch of failures because of changes to `copy` behaviour in `np.asarray`. I added a bullet point about this in https://github.com/apache/arrow/issues/39532. We might also need to wait until things are fixed for this in pandas to get a fully green build (https://github.com/pandas-dev/pandas/pull/57172)
